xline/src/server/watch_server.rs

```diff
@@ -0,0 +1,263 @@
+        }
+    }
+
+    /// Handle watch event
+    async fn handle_watch_event(&mut self, mut event: WatchEvent) {
+        let watch_id = event.watch_id();
+        let events = event.take_events();
+        let response = WatchResponse {
+            header: Some(ResponseHeader {
+                revision: event.revision(),
+                ..ResponseHeader::default()
+            }),
+            watch_id,
+            events,
+            ..WatchResponse::default()
+        };
+        assert!(
+            self.response_tx.send(Ok(response)).await.is_ok(),
+            "Receiver is closed"
+        );
+    }
+}
+
+#[tonic::async_trait]
+impl Watch for WatchServer {
+    ///Server streaming response type for the Watch method.
+    type WatchStream = ReceiverStream<Result<WatchResponse, tonic::Status>>;
+
+    /// Watch watches for events happening or that have happened. Both input and output
+    /// are streams; the input stream is for creating and canceling watchers and the output
+    /// stream sends events. One watch RPC can watch on multiple key ranges, streaming events
+    /// for several watches at once. The entire event history can be watched starting from the
+    /// last compaction revision.
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select!
+    async fn watch(
+        &self,
+        request: tonic::Request<tonic::Streaming<WatchRequest>>,
+    ) -> Result<tonic::Response<Self::WatchStream>, tonic::Status> {
+        debug!("Receive Watch Connection {:?}", request);
+        let req_stream = request.into_inner();
+        let (tx, rx) = mpsc::channel(128);
+        let watcher_clone = Arc::clone(&self.watcher);
+        let _hd = tokio::spawn(async move {
+            let (event_tx, event_rx) = mpsc::channel(128);
+            let mut watch_handle =
+                WatchHandle::new(watcher_clone, tx, req_stream, event_rx, event_tx);
+            loop {
+                tokio::select! {
+                    req = watch_handle.request_rx.next() => {
+                        if let Some(req) = req {
+                            match req {
+                                Ok(req) => {
+                                    watch_handle.handle_watch_request(req).await;
+                                }
+                                Err(e) => {
+                                    panic!("Receive WatchRequest error {:?}", e);
+                                }
+                            }
+                        } else {
+                            // TODO handle client close connection
+                            panic!("Client close connection");
+                        }
+                    }
+                    event = watch_handle.event_rx.recv() => {
+                        if let Some(event) = event {
+                            watch_handle.handle_watch_event(event).await;
+                        } else {
+                            panic!("Watch event sender is closed");
```

Comment:
Can we just jump out the loop and leave a log instead of panicing?

Reply:
So far it is a non-recoverable error as sender side is closed we cannot receive watch event any more. Will consider robustness in the future.